ephios/static/ephios/js/serviceworker.js

```diff
@@ -19,40 +8,86 @@ const filesToCache = [
 self.addEventListener('activate', event => {
     event.waitUntil(
         caches.keys().then(cacheNames => {
             return Promise.all(
                 cacheNames
-                    .filter(cacheName => (cacheName.startsWith("django-pwa-")))
+                    .filter(cacheName => (cacheName.startsWith("ephios-pwa-")))
                     .filter(cacheName => (cacheName !== staticCacheName))
                     .map(cacheName => caches.delete(cacheName))
             );
         })
     );
 });
 
-// Serve from Cache
+async function markResponseAsOffline(response) {
+    let content = await response.body.getReader().read().then(r => r.value);
+    let body = new TextDecoder().decode(content);
+    // this is somewhat hacky, but it works to communicate to the frontend that we are offline
+    body = body.replace("data-pwa-network=\"online\"", "data-pwa-network=\"offline\"");
+    return new Response(
+        new TextEncoder().encode(body),
+        {
+            headers: response.headers,
+            status: response.status,
+            statusText: response.statusText
+        }
+    );
+}
+
+async function cacheThenNetwork(event) {
+    // Return static files from the cache by default,
+    // falling back to network and caching it then.
+    let cache_response = await caches.match(event.request);
+    if (cache_response) {
+        return cache_response;
+    }
+    let response = await fetch(event.request);
+    let cache = await caches.open(staticCacheName);
+    await cache.put(event.request, response.clone());
+    return response;
+}
+
+async function networkThenCacheOrOffline(event) {
+    // Serve dynamic content from network, falling back to cache when offline.
+    // Cache network responses for the offline case.
+    try {
+        let response = await fetch(event.request);
+        if(!response.ok ) {
+            // fetch() on firefox does not throw an error when offline, but returns an Exception object
+            throw new Error("Response is not ok");
+        }
+        if (event.request.method === "GET" && response.status === 200) {
+            // This will inadvertently cache pages with messages in them
+            let cache = await caches.open(staticCacheName);
+            await cache.put(event.request, response.clone());
+        }
+        return response;
+    } catch (err) {
+        let response = await caches.match(event.request)
+        if (response) {
+            return await markResponseAsOffline(response);
+        }
+        return caches.match('/offline/', {ignoreVary: true});
+    }
+}
+
 self.addEventListener("fetch", event => {
-    event.respondWith(
-        caches.match(event.request)
-            .then(response => {
-                return response || fetch(event.request);
-            })
-            .catch(() => {
-                return caches.match('/offline/');
-            })
-    )
-});
+    const isStatic = new URL(event.request.url).pathname.startsWith("/static/");
```

Comment:
could use another way to descriminate between caching strategies:
![image](https://github.com/ephios-dev/ephios/assets/5970076/aa7630c8-2f3b-417b-9321-c55147bb5e05)
(from https://developer.chrome.com/docs/workbox/caching-strategies-overview/)